server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -83,7 +85,9 @@ public ConnectionManager(Settings settings, Transport transport, ThreadPool thre
     }
 
     public void addListener(TransportConnectionListener listener) {
-        this.connectionListener.listeners.add(listener);
+        if (connectionListener.listeners.contains(listener) == false) {
+            this.connectionListener.listeners.add(listener);
```

Comment:
as `connectionListener.listeners` is a `CopyOnWriteArrayList`, you can use `addIfAbsent`, which also has proper concurrency semantics

Reply:
I pushed https://github.com/elastic/elasticsearch/commit/b6fca90d9d96d3c311b2b607ebfa2deceb20d2c4